Devicetree: Devicetree Bindings: Support enums for array like dt props

JHirsbrunner opened the pull request:
It is currently impossible to use enum with any array like type (i.e. string-array and array, these are the only ones that make sense) in the devicetree and dt-bindings. However, there is no such remark in the dt-bindings section of the docs. Since this is a feature that comes in very handy and is implemented fairly easily, I adjusted the scripts for this.
- Update the edtlib.py and gen_defines.py scripts to support these types.
- Update enum.yaml, test.dts, and test_edtlib.py to test the newly added feature.
- Update devicetree.h to support the newly added feature.

It is now possible to do something like this.
```yaml
compatible = "enums"

properties:
  array-enum:
    type: string-array
    enum:
      - bar
      - foo
      - baz
      - zoo
```
```dts
/ {
	enums {
		compatible = "enums";
		array-enum = "foo", "bar";
	};
};
```

JHirsbrunner commented on scripts/dts/gen_defines.py:
```diff
@@ -732,6 +695,64 @@ def write_vanilla_props(node):
+    for i, subval in enumerate(val):
+        # DT_N_<node-id>_P_<prop-id>_IDX_<i>_ENUM_VAL_<val>_EXISTS 1
+        ret[macro + f"_IDX_{i}_ENUM_VAL_{subval}_EXISTS"] = 1
+        if not spec.enum_tokenizable:
+            continue
+
+        # DT_N_<node-id>_P_<prop-id>_IDX_<i>_ENUM_TOKEN
+        ret[macro + f"_IDX_{i}_ENUM_TOKEN"] = subval
+        if spec.enum_upper_tokenizable:
+            # DT_N_<node-id>_P_<prop-id>_IDX_<i>_ENUM_UPPER_TOKEN
+            ret[macro + f"_IDX_{i}_ENUM_UPPER_TOKEN"] = subval.upper()
```
why do they even exist, if there are already string tokens? Couldn't these few lines just be removed?

fgrandel replied:
Well observed. These are already deprecated, see devicetree.h, line 50ff. Maybe a good next PR for you to remove those everywhere? They've been deprecated for 3 years now, it seems. :-D

Thanks for factoring this code into methods, btw. Much more readable now.

decsny commented on include/zephyr/devicetree.h:
should update version of the api to 1.1.0

JHirsbrunner replied:
I guess the version needs to be updated again? Or is it fine to leave it at 1.1.0?

decsny commented on scripts/dts/python-devicetree/tests/test-bindings/enums.yaml:
```diff
@@ -32,5 +32,11 @@ properties:
       - 123 is ok
 
+  array-enum:                   # tokenizable string-array
+    type: string-array
```
please also add a test case for the `array` type

JHirsbrunner replied:
done

galak commented:
What's the actual use case for this?

JHirsbrunner commented:
> What's the actual use case for this?

On one side, there is no documentation that says you cannot use enumerations for string-arrays and arrays. On the other side, it is easily implemented with no side-effect.

And here how I stumbled across this issue:
There exists a register with some fields like this
- ASTOP[n:0]
- ASTART[m:0]

every bit within those fields stands for a different trigger source that performs the action (i. e. auto start or auto stop).
Since I don't want to create an header file with macros that the user has to include and use in an overlay or wherever, I wanted to use an enumerted string array.

JHirsbrunner commented:
As I was on vacation until 02.05.2024, I waited a week from then, before asking for another review.
@galak @decsny @henrikbrixandersen @gmarull What do you think? Can this, now that the checks pass, be merged?

decsny reviewed (approved):
well, I think it's good, but galak is the maintainer, and he appeared to have some reservation

JHirsbrunner commented:
@galak could you please review and give an opinion?
I answered your question a few weeks ago [here](https://github.com/zephyrproject-rtos/zephyr/pull/71497#issuecomment-2066650311).
More use cases could be pinctrl state names, where a developer could restrict the available options of the `pinctrl-names` property within a custom binding (just as an idea).

Thanks for your time.

decsny reviewed (dismissed):
since you are changing gen_defines.py, you also need to update doc/build/dts/macros.bnf

galak reviewed (dismissed):
Can you give an example of how this would get used to better understand the need.

If we go forward with this you need to update doc/build/dts/macros.bnf & tests//lib/devicetree/api/ for any new APIs added to devicetree.h

JHirsbrunner commented:
> Can you give an example of how this would get used to better understand the need.

Let's say we have a timer with a continuous feature. The timer can be started and stopped by trigger signals (TRG1, TRG2, TRG3). In addition to those triggers, it can also be stopped by period, and zero.

```.yaml
compatible: "vendor,chip-timer"

include: [base.yaml]

properties:
  < stuff here >

  auto-start:
    type: string-array
    description: |
      Sources for the auto start trigger of the timer.
    enum:
      - TRG1
      - TRG2
      - TRG3

  auto-stop:
    type: string-array
    description: |
      Sources for the auto stop trigger of the timer
    enum:
      - TRG1
      - TRG2
      - TRG3
      - ZERO
      - PER
```

What currently happens is as follows.
```.dts
/ {
    soc {
        tim1: timer@01234567 {
            compatible = "vendor,chip-timer";
            < stuff here >
            auto-start = "TRG1", "TRG2";
            auto-stop = "TRG3", "ZERO";
        };
    };
};
```
At the moment, this results in the error message "value of property 'auto-start' on {path} in {edt.dts_path} ([TRG1, TRG2]) is not in 'enum' list in {binding_path} ([TRG1, TRG2, TRG3])".

After the proposed change, these will compile:
```.dts
/ {
    soc {
        tim1: timer@01234567 {
            compatible = "vendor,chip-timer";
            < stuff here >
            auto-start = "TRG1", "TRG2";
            auto-stop = "TRG3", "ZERO";
        };
    };
};
```
```.dts
/ {
    soc {
        tim1: timer@01234567 {
            compatible = "vendor,chip-timer";
            < stuff here >
            auto-start = "TRG1";
            auto-stop = "ZERO", "ZERO";
        };
    };
};
```

And these will fail, since TRG5, foo, and bar are not part of the enums:
```.dts
/ {
    soc {
        tim1: timer@01234567 {
            compatible = "vendor,chip-timer";
            < stuff here >
            auto-start = "TRG1", "TRG5";
            auto-stop = "bar", "foo";
        };
    };
};
```
```.dts
/ {
    soc {
        tim1: timer@01234567 {
            compatible = "vendor,chip-timer";
            < stuff here >
            auto-start = "foo";
            auto-stop = "bar";
        };
    };
};
```
The error message of the first negative example is something along "value of property 'auto-start' on {path} in {edt.dts_path} (TRG5) is not in 'enum' list in {binding_path} ([TRG1, TRG2, TRG3])".

galak commented:
> Let's say we have a timer with a continuous feature. The timer can be started and stopped by trigger signals (TRG1, TRG2, TRG3). In addition to those triggers, it can also be stopped by period, and zero.

Thanks for the example, I'm curious if this is based on real world case, and if you'd just convert the string to some number in the code?

I'm concerned about the value prop here.  We have limited uses of string-array today, so trying to see the value.  I can see value in the validation checking side.  Trying to wrap my head still around the API side, and would be useful to see if you have an example of what the code side would look like.

jhedberg commented:
I'm happy to hear that this is being proposed. I would have needed exactly this when I was defining the new Bluetooth HCI base binding as part of #72323: https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/bluetooth/bt-hci.yaml

In particular, the `bt-hci-quirks` property is a string array that's only allowed to contain `BT_HCI_QUIRK_*` values from this enum: https://github.com/zephyrproject-rtos/zephyr/blob/d18fd46e991134ea29c8b52a52f2e4960327dfa4/include/zephyr/drivers/bluetooth.h#L38-L47

Example usage here (with just one array element, but you get the idea): https://github.com/zephyrproject-rtos/zephyr/blob/d18fd46e991134ea29c8b52a52f2e4960327dfa4/dts/bindings/bluetooth/zephyr%2Cbt-hci-ipc.yaml#L12-L13

I was able to limit the the possible values of `bt-hci-bus` through an enum, since it's just a single value, but even though the same would have been needed for the quirks that wasn't possible:
https://github.com/zephyrproject-rtos/zephyr/blob/d18fd46e991134ea29c8b52a52f2e4960327dfa4/dts/bindings/bluetooth/bt-hci.yaml#L9-L22

JHirsbrunner commented:
First of all, please excuse my absence over the past few weeks. I was bedridden with a headache.

> since you are changing gen_defines.py, you also need to update doc/build/dts/macros.bnf

> If we go forward with this you need to update doc/build/dts/macros.bnf & tests//lib/devicetree/api/ for any new APIs added to devicetree.h

I looked at `doc/build/dts/macros.bnf` and I don't think any changes are necessary; correct me if I'm wrong. The newly generated macros do not differ from the existing ones in that they use the same format and can be represented with the existing rules.
About the tests, I'll push a commit once I find the time to work on it (that will be this evening at the earliest).

> ... I'm curious if this is based on real world case, and if you'd just convert the string to some number in the code?

Yes, this is a simplified example of the Counters we use in an ASIC some of my colleagues designed. It might be niched, however, I thought it might be useful, even if only for the validation. The value of an string-array element can be represented three different ways. As a string, as a token and as an index value. An integer array can be represented two different ways. As the actual integer value, or as the index value of the actual integer value.

> I'm concerned about the value prop here. We have limited uses of string-array today, so trying to see the value. I can see value in the validation checking side. Trying to wrap my head still around the API side, and would be useful to see if you have an example of what the code side would look like.

In case of the example [above](https://github.com/zephyrproject-rtos/zephyr/pull/71497#issuecomment-2145543378)
```.dts
/ {
    soc {
        tim1: timer@01234567 {
            compatible = "vendor,chip-timer";
            reg = <0x01234567 0x100>;
            < stuff here >
            auto-start = "TRG1", "TRG2";
            auto-stop = "TRG3", "ZERO";
        };
    };
};
```
you can get the value of a string-array enum in three different ways with different representations.
```.c
/* expands to 0, 1 */
DT_FOREACH_PROP_ELEM_SEP(DT_NODELABEL(tim1), auto_start, DT_ENUM_IDX_BY_IDX, (, ))
/* expands to 2, 3 */
DT_FOREACH_PROP_ELEM_SEP(DT_NODELABEL(tim1), auto_stop, DT_ENUM_IDX_BY_IDX, (, ))

/* expands to TRG1, TRG2 */
DT_FOREACH_PROP_ELEM_SEP(DT_NODELABEL(tim1), auto_start, DT_STRING_UPPER_TOKEN_BY_IDX, (, ))
/* expands to TRG3, ZERO */
DT_FOREACH_PROP_ELEM_SEP(DT_NODELABEL(tim1), auto_stop, DT_STRING_UPPER_TOKEN_BY_IDX, (, ))

/* expands to "TRG1", "TRG2" */
DT_FOREACH_PROP_ELEM_SEP(DT_NODELABEL(tim1), auto_start, DT_PROP_BY_IDX, (, ))
/* expands to "TRG3", "ZERO" */
DT_FOREACH_PROP_ELEM_SEP(DT_NODELABEL(tim1), auto_stop, DT_PROP_BY_IDX, (, ))
```
This is the same behavior that already exists for non array enum properties.

On the other hand, in case of
```.yaml
properties:
  foo:
    type: array
    enum:
      - 3
      - 2
      - 1
      - 0
```
and
```.dts
&bar {
    foo = < 3 2 3 0 >;
};
```
```.c
/* expands to 0, 1, 0, 3 */
DT_FOREACH_PROP_ELEM_SEP(DT_NODELABEL(bar), foo, DT_ENUM_IDX_BY_IDX, (, ))
/* expands to 3, 2, 3, 0 */
DT_FOREACH_PROP_ELEM_SEP(DT_NODELABEL(bar), foo, DT_PROP_BY_IDX, (, ))
```

JHirsbrunner commented:
Here is another example, this time with a driver implementation in C. You could then use these enums as follows.
This still refers to this [example](https://github.com/zephyrproject-rtos/zephyr/pull/71497#issuecomment-2145543378)
```.dts
/ {
    soc {
        tim1: timer@01234567 {
            compatible = "vendor,chip-timer";
            reg = <0x01234567 0x100>;
            < stuff here >
            auto-start = "TRG1", "TRG2";
            auto-stop = "TRG3", "ZERO";
        };
    };
};
```
```.c
/* define the DT_DRV_COMPAT macro for DT_INST_ macros. */
#define DT_DRV_COMPAT vendor_chip_timer

union control_register {
    struct {
        uint32_t auto_start : 3;
        uint32_t auto_stop : 5;
        uint32_t : 24; /* other options */
    } bf;
    uint32_t value;
};
struct config {
    void *const base;
    union control_register cr;
    ...
};

#define CONFIG(dev) ((const struct device *)dev->config)
#define BASE(dev) ((union control_register *)(CONFIG(dev)->base))

static int init(const struct device *const dev)
{
    ...
    BASE(dev)->value = CONFIG(dev)->cr.value;
    ...
    return 0;
}
```
In the init function the control register of the timer is set to its initial state that can be configured in the devicetree. With these as the base, it is possible to initialize `cr` two different ways. 
The first possibility would be to use the indices of the enums.
```.c
#define DT_PROP_GET_BIT(node_id, prop, idx) BIT(DT_ENUM_IDX_BY_IDX(node_id, prop, idx))
#define DT_INST_SET_FIELD(inst, field) \
    .field = (DT_INST_FOREACH_PROP_ELEM_SEP(inst, field, DT_PROP_GET_BIT, (|)))
```

The second possibility would be to use the string tokens of these enums. Something among
```.c
#define TRG1 BIT(0)
#define TRG2 BIT(1)
#define TRG3 BIT(2)
#define ZERO BIT(3)
#define PER  BIT(4)

#define DT_INST_SET_FIELD(inst, field) \
    .field = (DT_INST_FOREACH_PROP_ELEM_SEP(inst, field, DT_STRING_UPPER_TOKEN_BY_IDX, (|)))
```

The rest of the code
```.c
#define DT_INST_SETUP(inst) \
    static const struct config config_##inst = { \
        .base = (void *)DT_INST_REG_ADDR(inst), \
        .cr = {.bf = {DT_INST_SET_FIELD(inst, auto_start), DT_INST_SET_FIELD(inst, auto_stop)}}, \
        ..., \
    }; \
    ... \
    DEVICE_DT_INST_DEFINE(inst, init, NULL, NULL, &config_##inst, POST_KERNEL, 50, &driver_api);

DT_INST_FOREACH_STATUS_OKAY(DT_INST_SETUP);
```

swift-tk commented:
Is there a way to assign values to enum in DTS? It seems like you are trying to have holes in between string enums?

JHirsbrunner commented:
> Is there a way to assign values to enum in DTS? It seems like you are trying to have holes in between string enums?

Sorry, I don't seem to understand your question. The value of a string enum is the string and the index is its position in the enum list. What do you mean with holes in between string enums? If you mean something like below, then the answer is no.
```.c
enum {
    FOO = 0,
    BAR = 20,
    ZOO = 10,
};
```
The idea is to get a string-array property work together with an enum restriction.
Enums for the dtc work like this:
```.dts
foo:
  type: string
  description: |
    0: opt1 < value is "opt1", index is 0
    1: opt2
    2: opt3
  enum:
    - opt1
    - opt2
    - opt3
```
It is - to my knowledge - not possible to make holes between the members of an enum.

swift-tk commented:
> The second possibility would be to use the string tokens of these enums. Something among
> 
> #define TRG1 BIT(0)
> #define TRG2 BIT(1)
> #define TRG3 BIT(2)
> #define ZERO BIT(3)
> #define PER  BIT(4)
> 
> #define DT_INST_SET_FIELD(inst, field) \
>     .field = (DT_INST_FOREACH_PROP_ELEM_SEP(inst, field, DT_STRING_UPPER_TOKEN_BY_IDX, (|)))
> https://github.com/zephyrproject-rtos/zephyr/blob/d18fd46e991134ea29c8b52a52f2e4960327dfa4/include/zephyr/drivers/bluetooth.h#L38-L47

It seems to me that real use case you and [jhedberg](https://github.com/jhedberg) mentioned are using enums with holes. Wouldn't it be much easier to work towards the direction so we could assign the BIT(x) or non sequential values in the yaml?

jhedberg commented:
> It seems to me that real use case you and [jhedberg](https://github.com/jhedberg) mentioned are using enums with holes. Wouldn't it be much easier to work towards the direction so we could assign the BIT(x) or non sequential values in the DTS?

I'm not sure I completely understand what you mean (an example would help), but ultimately what I'd want is to be able to keep using logical names in the binding and dts files, so that you don't need to go look up elsewhere what the values actually mean. The logical name doesn't necessarily have to be 1:1 with what the C code uses, i.e. existing enum values or preprocessor macros, but that seemed like the easiest way to implement the mapping today thanks to the existence of the `DT_STRING_TOKEN_*()` macros.

JHirsbrunner commented:
> It seems to me that real use case you and [jhedberg](https://github.com/jhedberg) mentioned are using enums with holes. Wouldn't it be much easier to work towards the direction so we could assign the BIT(x) or non sequential values in the DTS?

That's not the purpose of this PR. Yes, in my example I do use `BIT(x)` to assing values to the string upper token, however the use case is to use enum with a string / integer array. The idea is to have a validation check that works for arrays. Also the generated macros are built like any normal array and can be used with the enum macros.

Yes, it coincides with the value assignment in my case, however, I don't really care about that. You can just use the string token and assign a value with a macro expansion or c enum. However, it is currently not possible to use `type: string-array` together with `enum: ...` and then do stuff like `foo = "bar", "baz", "zoo";` in the devicetree, even if every option is within the enum specification, it will fail, saying `["bar", "baz", "zoo"] is not in [...]`.

JHirsbrunner commented:
@galak @decsny can this PR be merged now? All pipelines pass and there are only your requested changes open that I've taken care of.
Thanks for your time.

decsny commented:
> @galak @decsny can this PR be merged now? All pipelines pass and there are only your requested changes open that I've taken care of. Thanks for your time.

is a change to doc/build/dts/macros.bnf not needed? I havent got a chance to review the latest pr

JHirsbrunner commented:
> is a change to doc/build/dts/macros.bnf not needed?

I looked at `doc/build/dts/macros.bnf` and I don't think any changes are necessary; correct me if I'm wrong. The newly generated macros do not differ from the existing ones in that they use the same format and can be represented with the existing rules.

decsny commented:
I guess it seems like the macros.bnf might be missing some documentation related to strings and enums macros being generated related to a node, so it maybe needs updated but doesn't have to be done in this PR

JHirsbrunner commented:
@henrikbrixandersen @gmarull @galak How to proceed? Can we merge?

decsny reviewed (changes requested):
see above

fgrandel reviewed (commented):
@JHirsbrunner @jhedberg Sorry that I'm so late to the party but I got asked for a review just today.

A question to better understand your examples and use cases:

Your examples both seem to target bitfields (or could be modeled as such). We have been modeling similar use cases as macros included into the DT source in the past. Our Zephyr-specific DT parser is able to resolve quite complex arithmetic even.

Examples:
* Search for `NRF_PSEL` and `#define NRF_PSEL` or any of the other pinctrl abstractions.
* `<&gpiob 11 (GPIO_PULL_DOWN | GPIO_ACTIVE_HIGH)>,`

Why do we need this approach in addition to the existing one? Is it just an alternative way to do the same thing? Does it fill in a unique gap? Is it just a convenience? Is it to work around a badly modeled device tree property?

If it's the latter then maybe we should change the binding/DT definition from a string-array/array to an integer bitfield instead? If it's an alternative approach to an existing solution then I'd be rather sceptical as it would introduce additional modeling ambiguity into DT.

JHirsbrunner commented:
> Why do we need this approach in addition to the existing one? Is it just an alternative way to do the same thing? Does it fill in a unique gap? Is it just a convenience? Is it to work around a badly modeled device tree property?

As already mentioned, the behavior of string-arrays / arrays together with enums are undefined and, imho, a valid use-case.
It is true that bitfields could be resolved with macros (as in your example). However, when thinking about named properties, it might be useful to restrict options depending on the binding. Here an example with pinctrl.

```yaml
description: Some description
compatible: vend,some-unit
include: [base.yaml, pinctrl-device.yaml]

properties:
  pinctrl-0:
    required: true
  pinctrl-1:
    required: true
  pinctrl-2:
    required: true

  pinctrl-names:
    type: string-array
    required: true
    enum:
      - default
      - option1
      - option2
```
```.dtsi
&node: { // node with compat vend,some-unit
	pinctrl-0 = <setup for default>;
	pinctrl-1 = <setup for option2>; // doesn't depend on order
	pinctrl-2 = <setup for option1>;
	pinctrl-names = "default", "option1", "option2;
}
```

Note that there is no sleep in the options. Of course, you could add that as well. The idea is to simply limit the available options to an oversee-able spectrum that's also used in the c-code. Other use cases would be if you were to have any other named phandle-array and want to use the named property, meaning
```.dtsi
foos = <&phandle val1>, <&phandle val2>, <&phandle val3>;
foo-names = "option1", "option2", "option3";
```
If you write a driver or a module with a dt compat that looks up any value from `foos` by name, it is beneficial to limit the names to the ones actually used in the c-code (or vise versa). With this you'd get a compile error that is readable, in contrast to the printout of every macro expansion.

> If it's the latter then maybe we should change the binding/DT definition from a string-array/array to an integer bitfield instead? If it's an alternative approach to an existing solution then I'd be rather sceptical as it would introduce additional modeling ambiguity into DT.

If we don't want to support this, we should at the very least have it documented somewhere that array and string-array aren't supported for enum. The advantage of enum as string is that you can get the token, which you will need in the described example above. This should not make the modeling more ambiguous, but actually eliminate one of the ambiguities (undefined behaviors) that exist.

fgrandel commented:
@JHirsbrunner 

Thank you for taking all that time to answer me and giving such a detailed example. Be assured that your effort was not wasted. I read your comment carefully. +1 for this change from my side. I'll do a more detailed review then.

fgrandel commented on include/zephyr/devicetree.h:
```diff
@@ -858,44 +858,77 @@
- *     DT_ENUM_IDX(DT_NODELABEL(usb1), maximum_speed) // 1
- *     DT_ENUM_IDX(DT_NODELABEL(usb2), maximum_speed) // 3
+ *     DT_ENUM_IDX_BY_IDX(DT_NODELABEL(usb1), maximum_speed, 0) // 1
+ *     DT_ENUM_IDX_BY_IDX(DT_NODELABEL(usb2), maximum_speed, 0) // 3
```
(optional) nit - could you given an example of a nonzero index ideally (even more optional) including a snippet of DT that corresponds to the sample code? Similarly for the other macros, but won't mention it.

JHirsbrunner replied:
Well, I changed the current one...

fgrandel commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -765,13 +765,13 @@ class Property:
       Convenience for spec.type.
 
     val_as_token:
```
As you are breaking backwards compat of the signature anyway, you could also rename the member to "val_as_tokens" like you already did with "enum_indices".

JHirsbrunner replied:
I changed it

JHirsbrunner commented on include/zephyr/devicetree.h:
```diff
@@ -858,44 +858,77 @@
+ *         default_idx_value otherwise
+ */
+#define DT_ENUM_IDX_BY_IDX_OR(node_id, prop, idx, default_idx_value) \
+	COND_CODE_1(DT_NODE_HAS_PROP(node_id, DT_CAT3(prop, _IDX_, idx)), \
```
Of course, you're right. It slipped my attention, it's fixed now

JHirsbrunner commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -659,7 +659,7 @@ def enum(self) -> Optional[list]:
         "See the class docstring"
         if not hasattr(self, '_enum_tokenizable'):
-            if self.type != 'string' or self.enum is None:
+            if self.type not in {'string', 'string-array'} or self.enum is None:
```
Done

JHirsbrunner committed:
Devicetree: Devicetree Bindings: Support enums for array like dt props

It is currently impossible to use enum with any array like type (i.e.
string-array and array, these are the only ones that make sense) in the
devicetree and dt-bindings.
However, there is no such remark in the dt-bindings section of the docs.
Since this is a feature that comes in very handy and is implemented
fairly easily, I adjusted the scripts for this.

It is now possible to do something like this.
```yaml
compatible = "enums"

properties:
  array-enum:
    type: string-array
    enum:
      - bar
      - foo
      - baz
      - zoo
```
```dts
/ {
	enums {
		compatible = "enums";
		array-enum = "foo", "bar";
	};
};
```

Signed-off-by: Joel Hirsbrunner <jhirsbrunner@baumer.com>

JHirsbrunner committed:
Devicetree: Devicetree Bindings: Adjust python tests

Adjust existing tests to support the changes and add new tests to test
the newly added feature.

Signed-off-by: Joel Hirsbrunner <jhirsbrunner@baumer.com>

JHirsbrunner committed:
Devicetree: Devicetree Bindings: Add new APIs

The existing APIs for enums were rewritten to support the changes.
Additional macros were added to now also support getting specific
indices from the array / string-array.
The macros can be evoked the same way like before, since nothing changed
for the interface. Therefore, these changes are backward compatible.

Signed-off-by: Joel Hirsbrunner <jhirsbrunner@baumer.com>

JHirsbrunner committed:
Devicetree: Devicetree Bindings: Add tests for new DT_ENUM_ macros

Test the new c-macros for dt enums. The new macros are already used in
the existing macros. As an example, DT_ENUM_IDX(node_id, prop) uses
DT_ENUM_IDX_BY_IDX(node_id, prop, 0) to get its result. However, this is
insufficient for testing the complete functionality of these macros.
Therefore, additional tests are added to make sure they work
appropriately for other indices besides 0.

Signed-off-by: Joel Hirsbrunner <jhirsbrunner@baumer.com>

JHirsbrunner reviewed (commented):
I split the first commit into three:
- The first for changes in the python script
- Second for python script tests
- Third for changes to the c-headers

This should make it much more understandable.

fgrandel commented:
@galak Still objections? Maybe you want to refresh your review, since changes have been applied to this PR and your vote seems to be the last one blocking this PR...?

tmon-nordic commented:
This just merged while I was reviewing. I independently observed the need for string-array with enums (and even came with some bare-bones implementation because I didn't know about this PR). See #79634 for a problem description (and implementation) where string-array with enum seem to be the best solution.

fgrandel commented:
> This just merged while I was reviewing. I independently observed the need for string-array with enums (and even came with some bare-bones implementation because I didn't know about this PR). See #79634 for a problem description (and implementation) where string-array with enum seem to be the best solution.

Not sure - does your comment mean "good that it was contributed" or does it mean that you see a problem and that your proposed solution would have been superior for some reason?

tmon-nordic commented:
> Not sure - does your comment mean "good that it was contributed" or does it mean that you see a problem and that your proposed solution would have been superior for some reason?

It means "excellent work, thanks for doing that". The implementation here is better than mine. I just removed my implementation and all the work on top of that worked as expected.
